force-app/main/default/classes/DMLOperations.cls

```diff
@@ -181,14 +241,22 @@ public with sharing class DMLOperations {
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
 		// Loop through each Contact
-
+        for (Contact contactRecord : contactsToUpsert) {
 		// Extract the Account Name from Contact's LastName
+            Account[] contactAccount = [Select Id from Account where Name = :contactRecord.LastName];
 
 		// Get the account based on the name or create one if it doesn't exist
+            if(contactAccount.size() == 0 ){
+                contactAccount = new Account[1];
+                contactAccount[0] = upsertAccount(contactRecord.LastName);
+            }
 
-		// Link the Contact to its Account using the Account Id
+        // Link the Contact to its Account using the Account Id        
+            contactRecord.AccountId = contactAccount[0].Id;
+        }
 
 		// Upsert the Contacts
+        upsert contactsToUpsert;
 	}
```

Comment:
I would be easier to use the upsertAccounts method. But both solutions have DML etc in a loop. This is good for learning.